src/cargo/util/lints.rs

```diff
@@ -79,23 +86,37 @@ pub struct Lint {
 
 impl Lint {
     pub fn level(&self, lints: &TomlToolLints, edition: Edition) -> LintLevel {
+        let edition_level = self
+            .edition_lint_opts
+            .filter(|(e, _)| edition >= *e)
+            .map(|(_, l)| l);
+
+        if self.default_level == LintLevel::Forbid || edition_level == Some(LintLevel::Forbid) {
+            return LintLevel::Forbid;
+        }
+
         let level = self
             .groups
             .iter()
             .map(|g| g.name)
             .chain(std::iter::once(self.name))
             .filter_map(|n| lints.get(n).map(|l| (n, l)))
-            .max_by_key(|(n, l)| (l.priority(), std::cmp::Reverse(*n)));
+            .max_by_key(|(n, l)| {
+                (
+                    l.level() == TomlLintLevel::Forbid,
+                    l.priority(),
+                    std::cmp::Reverse(*n),
+                )
+            });
 
         match level {
             Some((_, toml_lint)) => toml_lint.level().into(),
             None => {
-                if let Some((lint_edition, lint_level)) = self.edition_lint_opts {
-                    if edition >= lint_edition {
-                        return lint_level;
-                    }
+                if let Some(level) = edition_level {
+                    level
+                } else {
+                    self.default_level
                 }
-                self.default_level
             }
```

Comment:
The implementation for default and edition exists in two places.  We should do this in one place.